### ISSUE TITLE
Return more than the username, password and One-Time Password (if available) in the login dictionary

### ISSUE THREAD

**Issue by vizh** [agilebits/onepassword-app-extension#211]
Is it possible to return website url or hostname in login dictionary keys result after using 1Password extension?

**Comment by radazzouz** [comment]
Hi @ViZH,

Thanks for taking the time to write in :+1: 

No it is not. `findLoginForUrlString:` requires the URL string as a parameter so there is no need to return it in the completion handler. Just out of curiosity, why would you like to have the item's URL(s) returned?

Closing this issue for now!

Thank you!

**Comment by victor** [comment]
Hi,

 I'd like to re-open this request. Our user case is as follows: We have a web application where every customer logs in through a dedicated sub-domain (i.e customer1.webapp.com).
When logging in via the web, the account that you want to use is obvious, because it is in the URL. But on the corresponding iOS app, we have the users fill a third field, which is the account name.

 I'm now integrating the 1P extension and the problem is that, since I do not know beforehand which account to use, I just request the top leve domain, "webapp.com". This finds all the login items that match the URL, including subdomains. This is great because it is possible, at least for us the developers, to have more than one account. But even if it is only one account that matches, or even if there is more than one account but all belonging to the same subdomain, only the username and password are filled. The account field cannot be filled because I lack the information to do it. Having the matched URL would be great. Alternatively, having access to the web form details would also be useful, because we could use a third field in the website to pre-fill the account name.

**TL;DR:** Since `findLoginForUrlString:` can match subdomains, it would be useful to know which one the user chose, as it may be necessary info for completing the authentication.

**Comment by radazzouz** [comment]
Hi @victor,

Thanks so much for staying your use-case :+1:

I appreciate your need to obtain more than just the credential information (username, password and OTP if available) via the login dictionary. However, for now we only want to put login information in the dictionary when using `findLoginForUrlString:`.

So I am re-opening this issue and labeling it with `under consideration`, hoping to get more votes from the community. We treat feature requests based on popularity and votes.

In the meantime, may I suggest the following user experience for your app?

* Opt for a multipage login.
* In the first screen prompt the user to type in the account name.
* The invoke the 1Password App Extension to fill in the username and password.

Hope that this helps :smiley:

**Comment by victor** [comment]
HI @radazzouz, I appreciate your suggestion. For now we'll have the third field there in the same screen, so that they can fill it manually, but in one go. I think that's better UX.

**Comment by radazzouz** [comment]
Thanks so much for the quick response, @victor :+1: 

That makes sense too. I guess I was under the influence of Slack which uses a separate screen for the team subdomain :wink: 

Also, when you're done adding 1Password into your app, please give us shout at https://blog.agilebits.com/contact-developers/ to submit it to our ["Apps ❤ 1Password" page](https://blog.agilebits.com/1password-apps/).

Cheers!

**Comment by coryalder** [comment]
Hello! Just wanted to throw in a +1 for this. Developing an app with exactly the same issue, multiple subdomains. It'd be great to be able to get the AppExtensionURLStringKey back.

:)

**Comment by radazzouz** [comment]
Thanks for the feedback @coryalder! :+1:

**Comment by dmytton** [comment]
+1 for this. Being able to authenticate in a single step is a smoother UX.

**Comment by radazzouz** [comment]
Glad you like it @dmytton! 😊

**Comment by jpgoldberg** [comment]
Hmm. I definitely see why app developers want this.

The question is whether we should give to the app the same sort of thing that a web server would get (basically everything in the URL).

The other question is what to do with when there are multiple websites for a Login. I do not want to provide all of them, as the user, when doing a "go and fill", would be the one picking which URL to use.

**Comment by atika** [comment]
+1 For my use-case I stock the services API Key into each account identifier item (connexion) in a password. As some services use this API Key to login like a password, and It was not possible to retrieve others fields that the password, I had to create another item, for each account, just to stock the API Key as password.

**Comment by lithium3141** [comment]
Just want to throw my support in for this feature as well. My company has a couple apps that offer sync over WebDAV, and one of the options our customers have is to provide their own WebDAV server (or use a third-party offering). In that case, the setup flow involves typing in the entire WebDAV URL, not just a team name or subdomain as in Slack.

As a complete URL, this sync location is potentially the most complicated piece of data attached to the account – even more so than the password. It would be especially handy in this setup flow if 1Password could return the URL attached to the Login chosen by the user.

As to the issue of multiple URLs: I'd think apps could implement a post-callback picker to allow the user to choose the desired URL if necessary. Alternately, 1Password could choose a single "canonical" URL for a given Login (the first one? explicitly designated? chosen in-extension?) and return only that URL.

Finally, as to the format: the critical bits of the URL in our case are the protocol, host, port, and path. We don't necessarily need in-URL username info (`lithium3141@example.com`), nor do we need query parameters or the fragment. Again, this is something we could postprocess out in the completion block if necessary; 1Password doesn't need to do this filtering on the app's behalf.

Thanks for being open to this kind of feedback, @radazzouz and @jpgoldberg! It's been a pleasure to work with the 1Password extension so far, and the AgileBits team has been wonderful about interacting with the community. 👍

**Comment by Jipsica** [comment]
I would like to support this feature as well, we also require multiple subdomains for login and this would definitely improve the flow.

I'm also wondering if it would be possible to return multiple one-time passwords in the dictionary if they exist. The reason being, if there are multiple subdomains for the login, then it's possible to have multiple one-time passwords associated as well. Right now there is no way to match up the correct one-time password to the selected subdomain. The dictionary only returns the first one-time password associated with the login.

This means the user still needs to exit our app, copy the correct one-time password for two-factor authentication, open our app, complete login. This is cumbersome for the user and the UI provides no information on which code to use!

<img width="451" alt="screen shot 2016-11-24 at 10 22 33 am" src="https://cloud.githubusercontent.com/assets/22913771/20603550/05bc63ce-b230-11e6-8f7d-d6057f947d51.png">

**Comment by imuller** [comment]
+1 for this feature, we need this 👍

**Comment by weien** [comment]
+1 for this feature! We have a use-case that's very similar to what @victor described.

**Comment by walteryaron** [comment]
+1 for this feature

**Comment by imuller** [comment]
Any update on this?

**Comment by gks** [comment]
@imuller No changes to the above. It's still under consideration but not something being actively worked on at this time.

**Comment by localcelebrity** [comment]
+1 for this. ActiveCampaign's login uses three fields: Account, Username, & Password. Our mobile app integrates with 1PW and this would dramatically improve the user experience for our customers.

**Comment by rudyrichter** [comment]
In light of the retirement of this extension, I'm closing out any remaining issues and branches.
